app/src/main/java/com/locked/shingranicommunity/auth/RegisterViewModel.kt

```diff
@@ -20,15 +24,82 @@ class RegisterViewModel @Inject constructor(
     val message: LiveData<String> = data.message
     val loading: LiveData<Boolean> = data.loading
 
-    fun setEmail(email: String) { data.email.postValue(email) }
-    fun setName(name: String) { data.name.postValue(name) }
-    fun setPassword(password: String) { data.password.postValue(password) }
-    fun setPasswordConfirm(passwordConfirm: String) { data.passwordConfirm.postValue(passwordConfirm) }
+    private val registerStateObserver: Observer<UserRepository.RegisterData> = Observer<UserRepository.RegisterData> {
+        it?.let {
+            if (it.registerSuccess) {
+                navigation.navigateToLogin(false)
+            } else {
+                data.loading.postValue(false)
+                data.message.postValue(it.registerMessage)
+            }
+        }
+    }
+    init {
+        userRepository.registerState.observeForever(registerStateObserver)
+    }
+
+    fun setEmail(email: String) {
+        data.email.postValue(email)
+        isUserNameValid(email)
+    }
+    fun setName(name: String) {
+        data.name.postValue(name)
+    }
+    fun setPassword(password: String) {
+        data.password.postValue(password)
+    }
+    fun setPasswordConfirm(passwordConfirm: String) {
+        data.passwordConfirm.postValue(passwordConfirm)
+    }
     fun onRegisterPress() {
-        data.loading.postValue(true)
-        data.message.postValue(resourceProvider.getString(R.string.please_wait))
-        // todo
+        when {
+            isUserNameValid(data.name.value!!) && validateEmail(data.email.value!!) && validatePassword(data.password.value!!, data.passwordConfirm.value!!) -> {
+                data.loading.postValue(true)
+                data.email.value?.let { email -> data.password.value?.let { password -> data.name.value?.let { name -> userRepository.register(email, password, name) } } }
+            }
+            else -> {
+                data.loading.postValue(false)
+            }
+        }
```

Comment:
This can be simplified:
```kotlin
 when {
        validateName(data.name.value!!) && validateEmail(data.email.value!!) && validatePassword(data.password.value!!, data.passwordConfirm.value!!) -> {
               userRepository.register(email, password, name)
     } else -> {
                data.loading.postValue(false)
     }
```